File: src/main/scala/llmTests/a201401b/scala/gpt35turbo/ProgressiveAcceptorImpl.scala
package llmTests.a201401b.scala.gpt35turbo

import llmTests.a201401b.scala.{Aggregator, ProgressiveAcceptor, ProgressiveFilter}

class ProgressiveAcceptorImpl[X] extends ProgressiveAcceptor[X] {

  private var filter: ProgressiveFilter[X] = _
  private var aggregator: Aggregator[X] = _
  private var size: Int = _
  private var sequence: Seq[X] = Seq.empty

  override def setProgressiveFilter(filter: ProgressiveFilter[X]): Unit = {
    this.filter = filter
  }

  override def setAggregator(aggregator: Aggregator[X]): Unit = {
    this.aggregator = aggregator
  }

  override def setSize(size: Int): Unit = {
    if (size <= 0) {
      throw new IllegalArgumentException("Not negative!")
    }
    this.size = size
  }

  override def accept(pos: Int, elem: X): Boolean = {
    if (filter == null || aggregator == null || size <= 0) {
      throw new IllegalStateException("Not correctly initialized")
    }

    if (pos < 0 || pos > size) {
      throw new IndexOutOfBoundsException("Invalid position")
    }

    val isAccepted = if (pos == 0 || filter.isNextOK(sequence(pos - 1), elem)) {
      if (pos <= sequence.length) {
        sequence = sequence.take(pos)
      }
      sequence = sequence :+ elem
      true
    } else {
      false
    }

    if (sequence.length > size) {
      sequence = sequence.takeRight(size)
    }

    isAccepted
  }

  override def aggregateAll(): X = {
    if (aggregator == null || sequence.isEmpty) {
      throw new IllegalStateException("Not correctly initialized")
    }

    sequence.reduceOption(aggregator.aggregate).getOrElse(throw new IllegalStateException("Empty sequence"))
  }
}

File: src/main/scala/llmTests/a201401b/java/gpt35turbo/ProgressiveAcceptorImpl.java
package llmTests.a201401b.java.gpt35turbo;

import llmTests.a201401b.java.Aggregator;
import llmTests.a201401b.java.ProgressiveAcceptor;
import llmTests.a201401b.java.ProgressiveFilter;

import java.util.ArrayList;
import java.util.List;

public class ProgressiveAcceptorImpl<X> implements ProgressiveAcceptor<X> {

    private List<X> sequence;
    private ProgressiveFilter<X> filter;
    private Aggregator<X> aggregator;
    private int size;

    public ProgressiveAcceptorImpl() {
        this.sequence = new ArrayList<>();
    }

    @Override
    public void setProgressiveFilter(ProgressiveFilter<X> filter) {
        this.filter = filter;
    }

    @Override
    public void setAggregator(Aggregator<X> aggregator) {
        this.aggregator = aggregator;
    }

    @Override
    public void setSize(int size) {
        if (size < 0) {
            throw new IllegalArgumentException("Size must be non-negative");
        }
        this.size = size;
    }

    @Override
    public boolean accept(int pos, X elem) {
        if (filter == null || aggregator == null || size == 0) {
            throw new IllegalStateException();
        }
        if (pos == 0) {
            sequence.clear();
        }
        boolean isNextOK = true;
        if (pos == sequence.size()) {
            if (pos >= size) {
                sequence.remove(0);
                pos--;
            }
            for (int i = 0; i < sequence.size(); i++) {
                X previous = sequence.get(i);
                if (!filter.isNextOK(previous, elem)) {
                    isNextOK = false;
                }
            }
            if (isNextOK) {
                sequence.add(elem);
            }
        } else if (pos < sequence.size()) {
            for (int i = pos; i < sequence.size(); i++) {
                sequence.remove(i);
            }
            accept(pos, elem);
        } else {
            throw new IllegalArgumentException();
        }
        return isNextOK;
    }

    @Override
    public X aggregateAll() {
        X result = null;
        if (sequence.size() > 0) {
            result = sequence.get(0);
            for (int i = 1; i < sequence.size(); i++) {
                X next = sequence.get(i);
                result = aggregator.aggregate(result, next);
            }
        }
        return result;
    }
}
